llvm/lib/CAS/OnDiskGraphDB.cpp

```diff
@@ -1310,3 +1342,111 @@ OnDiskGraphDB::OnDiskGraphDB(StringRef RootPath, OnDiskHashMappedTrie Index,
 OnDiskGraphDB::~OnDiskGraphDB() {
   delete static_cast<StandaloneDataMapTy *>(StandaloneData);
 }
+
+Error OnDiskGraphDB::importFullTree(ObjectID PrimaryID,
+                                    ObjectHandle UpstreamNode) {
+  // Copies the full CAS tree from upstream. Uses depth-first copying to protect
+  // against the process dying during importing and leaving the database with an
+  // incomplete tree. Note that if the upstream has missing nodes then the tree
```

Comment:
Any reason why this decision is made? I might go the other way and consider this an error but check `UpstreamDB` has the same fault-in policy as current DB?

Reply:
> Any reason why this decision is made?

It is important to allow flexibility for the client (or the higher-level API) to decide whether it wants to create lazy nodes or not, I don't think it should be something that `OnDiskGraphDB` itself should force.

> check UpstreamDB has the same fault-in policy as current DB?

This is already enforced, there will be an error if you try to open an `OnDiskGraphDB` with different fault-in policy than intended. It's technically not possible to chain two `OnDiskGraphDB` instances with different policy.

Reply:
> It's technically not possible to chain two OnDiskGraphDB instances with different policy.

Actually the client can do that if they explicitly opened them with different policies and then chained them. But as I mentioned I'd like to lean on having these "low-level building blocks" providing flexibility, the client can choose to use them in a more restrictive manner if it wants to (e.g. the client can only open and chains DBs with the same policy only).